### PR TITLE
customization env vars

### PR DESCRIPTION
**Pull request by fmigneault**
Retrieve env variable updates from #114 for separate PR
Removed the specific definitions to weaver, only 'server' related variables are preserved. 

Change overview:
* allow to customize `/data` persistence root on disk, retaining current default for existing deployment
* add data persistence for `mongodb` container

### CODE REVIEW

**Review by dbyrns — APPROVED** [review]
LGTM

File: birdhouse/scripts/backup-jupyterhub-notebooks.sh [hunk 1]
```diff
@@ -6,7 +6,7 @@ if [ -z "$BACKUP_OUT_DIR" ]; then
 fi
 
 if [ -z "$JUPYTERHUB_USER_DATA_DIR" ]; then
-    JUPYTERHUB_USER_DATA_DIR=/data/jupyterhub_user_data
+    JUPYTERHUB_USER_DATA_DIR=${DATA_PERSIST_ROOT:-/data}/jupyterhub_user_data
```
Comment (tlvu) [inline birdhouse/scripts/backup-jupyterhub-notebooks.sh:9]: Can't remember, but if this script do not source `defualt.env` and/or `env.local` then yes you should duplicate the default value here.  Otherwise please avoid duplication when possible.

File: birdhouse/scripts/bootstrap-testdata [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 
 if [ -z "$DATASET_ROOT" ]; then
-    DATASET_ROOT="/data/datasets"
+    DATASET_ROOT="${DATA_PERSIST_ROOT:-/data}/datasets"
```
Comment (tlvu) [inline birdhouse/scripts/bootstrap-testdata:10]: Can't remember, but if this script do not source defualt.env and/or env.local then yes you should duplicate the default value here. Otherwise please avoid duplication when possible.

**Review by tlvu — CHANGES_REQUESTED** [review]
Please review comments below and try to get a working build or assess build failure are unrelated.

File: birdhouse/components/scheduler/config.yml.template [hunk 3]
```diff
@@ -6,7 +6,7 @@
   dockerargs: >-
     --rm --name logrotate
     --volume /var/log/PAVICS:/var/log/PAVICS:rw
-    --volume /data/logrotate:/var/lib:rw
+    --volume ${DATA_PERSIST_ROOT:-/data}/logrotate:/var/lib:rw
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/components/scheduler/config.yml.template:9]: agreed

File: birdhouse/components/scheduler/deploy_data_job.env [hunk 4]
```diff
@@ -33,7 +33,7 @@ fi
 # Location for local cache of git clone to save bandwidth and time from always
 # re-cloning from scratch.
 if [ -z "$DEPLOY_DATA_JOB_CHECKOUT_CACHE" ]; then
-    DEPLOY_DATA_JOB_CHECKOUT_CACHE="/data/deploy_data_cache/$DEPLOY_DATA_JOB_JOB_NAME"
+    DEPLOY_DATA_JOB_CHECKOUT_CACHE="${DATA_PERSIST_ROOT:-/data}/deploy_data_cache/${DEPLOY_DATA_JOB_JOB_NAME}"
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/components/scheduler/deploy_data_job.env:36]: For this one, I'm not too sure what will happen since it is not a template file that would be processed by `pavics-compose`. Will it actually find the value from `default.env`? If so, indeed doesn't need the default again.

File: birdhouse/config/canarie-api/docker_configuration.py.template [hunk 5]
```diff
@@ -256,8 +256,8 @@ PLATFORMS = {
         'redirect': {
             'doc': 'https://pavics-sdi.readthedocs.io/en/latest/arch/backend.html',
             'releasenotes': 'https://github.com/bird-house/birdhouse-deploy/releases',
-            'support': 'https://github.com/Ouranosinc/pavics-sdi/issues',
-            'source': 'https://github.com/Ouranosinc/pavics-sdi',
+            'support': '${SUPPORT_URL}',
+            'source': 'https://github.com/bird-house/birdhouse-deploy',
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/config/canarie-api/docker_configuration.py.template:260]: I would like the feedback also. Maybe @dbyrns also has some insights on that. 
Feels weird that the "source" of the platform points to the docs repo though.

File: birdhouse/pavics-compose.sh [hunk 6]
```diff
@@ -25,10 +26,17 @@ VARS='
   $POSTGRES_PAVICS_PASSWORD
   $POSTGRES_MAGPIE_USERNAME
   $POSTGRES_MAGPIE_PASSWORD
+  $SERVER_NAME
+  $SERVER_SYNOPSIS
+  $SERVER_INSTITUTION
+  $SUPPORT_EMAIL
+  $SUPPORT_URL
+  $DATA_PERSIST_ROOT
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/pavics-compose.sh:34]: Well that's the point, I don't want to force it. You can override if you want. If not, leave them as is and default are used.
But if they don't exist at all (as it can be the case when in `OPTIONAL_VARS`), they will break the platform or cause missing key, URL, etc. errors.

File: birdhouse/pavics-compose.sh [hunk 7]
```diff
@@ -25,10 +26,17 @@ VARS='
...
 '
 
-# list of vars to be substituted in template but they do not have to be set in
-# env.local
+# list of vars to be substituted in template but they do not have to be set in env.local
+#   these variables can be omitted completely as they can be inferred from other values or are not specifically required
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/pavics-compose.sh:38]: I can reword it as suggested.

File: birdhouse/scripts/backup-jupyterhub-notebooks.sh (hunk 1 above)
Comment (fmigneault) [reply birdhouse/scripts/backup-jupyterhub-notebooks.sh:9]: I'm not sure either because its not a .template. 
That's why I don't take the chance.

File: birdhouse/scripts/bootstrap-testdata (hunk 2 above)
Comment (fmigneault) [reply birdhouse/scripts/bootstrap-testdata:10]: idem as other

File: birdhouse/docker-compose.yml [hunk 8]
```diff
@@ -383,6 +387,7 @@ networks:
 
 volumes:
   data: {}
+  mongodb: {}
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/docker-compose.yml:390]: This avoids missing definition of volume name. 
Could be used by any components, for example by weaver, but not limited to weaver specifically.

File: birdhouse/default.env [hunk 9]
```diff
@@ -1,15 +1,27 @@
 # All env in this default.env can be overridden by env.local.
 # All env in this default.env must not depend on any env in env.local.
 
+# General metadata details about the server instance
+# following are the information that will be provided by Canarie-API to describe the server
+# custom values are recommended for your production server
+export SERVER_NAME="PAVICS"
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/default.env:7]: will make it lowercase

File: birdhouse/default.env [hunk 10]
```diff
@@ -1,15 +1,27 @@
...
+export SERVER_SYNOPSIS="The PAVICS (Power Analytics for Visualization of Climate Science) platform is a collection of climate analysis services served through Open Geospatial Consortium (OGC) protocols. These services include data access, processing and visualization. Both data and algorithms can be accessed either programmatically, through OGC-compliant clients such as QGIS or ArcGIS, or a custom web interface."
+export SERVER_INSTITUTION="Demo Server"
+export SUPPORT_URL=https://ouranosinc.github.io/pavics-sdi/support/index.html
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/default.env:10]: I think this one aligns with the other comment about the source. 
If it really is for the organization, it should be override-able. 
For bw compat, should be a custom override for your prod, not something for the shared repo.

File: birdhouse/docker-compose.yml [hunk 11]
```diff
@@ -266,8 +266,12 @@ services:
   mongodb:
     image: mongo:3.4.0
     container_name: mongodb
+    volumes:
+      - ${DATA_PERSIST_ROOT}/mongodb_persist:/data/db
     # Mongodb crash with permission denied errors if the command is not overriden like this
     command: bash -c 'chown -R mongodb:mongodb /data && chmod -R 755 /data && mongod'
+    networks:
+      - mongodb
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/docker-compose.yml:274]: Yes, unless they are also in the mongodb network.

File: birdhouse/config/canarie-api/docker_configuration.py.template (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbyrns) [reply birdhouse/config/canarie-api/docker_configuration.py.template:260]: The section in question is "platform" which mean the frontend for Canarie. It used to be the react frontend, but now the notebooks are really the frontend, so pavics-sdi make sense I think. @fmigneault, if you look above you'll see a node section which is the backend and is correctly using birdhouse-deploy as source.

[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/config/canarie-api/docker_configuration.py.template:260]: I see. In that case, I will revert both of above `support` and `source` to the original values. 
Does it also apply to `PLATFORMS['pavics']['info']` elements as well? 

How should another organization define their configuration? 
The way I see it now, it's that every server will identify itself as the same `pavics` platform definition. 
Is this valid considering any organization can customize it with different features enabled/disabled?
In other words, is that definition still valid as the *default* template shared and not organization-specific?

[the remark this answers is not in the document]
Comment (dbyrns) [reply birdhouse/config/canarie-api/docker_configuration.py.template:260]: > Does it also apply to PLATFORMS['pavics']['info'] elements as well?

Yes, PLATFORMS['pavics'] apply to the frontend, while SERVICES['node'|'indices'|...] apply to backend core or services

> How should another organization define their configuration?

I think @tlvu is using the canarie api for monitoring with the an optional component (https://github.com/bird-house/birdhouse-deploy/tree/master/birdhouse/optional-components/canarie-api-full-monitoring), but for other organization, the canarie API itself is completely irrelevant. A standard nginx could be used there and move the whole canarie-api image as an optional component (https://github.com/bird-house/birdhouse-deploy/blob/master/birdhouse/docker-compose.yml#L4)

Anyway, this is out of scope right now and this whole config could be empty for someone not requiring it.

File: birdhouse/pavics-compose.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/pavics-compose.sh]: I re-read your comment. 
> why not just say "their default values are from default.env so they do not have to be defined in env.local".

and 
> is misleading [...] Because they are required for real during template expansion

Both are incorrect. They are not defined in `default.env`, nor are *required* anywhere.
And, they really are variables that are computed one way or another from other values, which can be overridden for convenience. 

Best example, first one `PAVICS_FQDN_PUBLIC`, is defined only in `pavics-compose.sh` and, uses `PAVICS_FQDN` if not defined, and even uses another value in `configure-pavics.sh` when using Kite.

File: birdhouse/config/canarie-api/docker_configuration.py.template (hunk 5 above)
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/config/canarie-api/docker_configuration.py.template]: I reverted all env-var metadata utils I initially added because it seems too much trouble to change without blocking this PR... 
Opened another issue #159.

I still strongly believe that this config is plain wrong. 
It makes senses only when deployed to `pavics.ouranos.ca`, and all other instances report erroneous references. 
But since there is much opposition to proposed changes, until then, it looks like this will remain the prod config of Ouranos server.

File: birdhouse/components/scheduler/deploy_data_job.env (hunk 4 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/components/scheduler/deploy_data_job.env:36]: This `deploy_data_job.env ` is sourced in `env.local` https://github.com/bird-house/birdhouse-deploy/blob/02d1e9db0cfc0ca0c2ca382e379228a48d4ca182/birdhouse/env.local.example#L180 and `env.local` should always be read after `default.env` so should not need to repeat the default again.

File: birdhouse/config/canarie-api/docker_configuration.py.template (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbyrns) [reply birdhouse/config/canarie-api/docker_configuration.py.template]: In fact, like you said, this whole config is so much linked to Ouranos and because other organization could have a completely different setup (using optional components), the easiest way would be to force each deployment to define its own config file and set the location in an environment variable here : https://github.com/bird-house/birdhouse-deploy/blob/master/birdhouse/docker-compose.yml#L22. The current template file would then be turned into an example file.

[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/config/canarie-api/docker_configuration.py.template]: > I think @tlvu is using the canarie api for monitoring with the an optional component (https://github.com/bird-house/birdhouse-deploy/tree/master/birdhouse/optional-components/canarie-api-full-monitoring),

Exact.

>  but for other organization, the canarie API itself is completely irrelevant. 

Exact, but they can also leverage the monitoring side, like Ouranos does.

All other "description/information" links are indeed completely irrelevant so they can be "wrong" without impact.

> A standard nginx could be used there and move the whole canarie-api image as an optional component (https://github.com/bird-house/birdhouse-deploy/blob/master/birdhouse/docker-compose.yml#L4)

Agreed.  Currently, nginx and the canarie monitoring cronjob and web interface https://github.com/Ouranosinc/CanarieAPI/blob/73835095262093b2c67e14a0659e8874b5720794/Dockerfile#L32-L35 are all running in the same `proxy` container.

With the current optional component mechanism, they can be splitted.

File: birdhouse/default.env (hunk 10 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/default.env]: Just to be clear why I am confused.

If you think this shared repo should not have something Ouranos specific so why the new suggested value is still pavics-sdi?

If on the contrary, if having pavics-sdi here is fine, since it can be overridden, why change the value for another pavics-sdi link?  Why not just keep the same?

File: birdhouse/docker-compose.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/docker-compose.yml:390]: Why add a mongodb volume that is not even used by any containers (not even mongodb itself)?

How can you have "missing definition of volume name" if the volume is not even used anywhere in this PR?

File: birdhouse/pavics-compose.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/pavics-compose.sh]: > Both are incorrect. They are not defined in `default.env`, nor are _required_ anywhere.

So there are only 3 exceptions that are not in `default.env` and the default value computed in `pavics-compose.sh` directly.  They are not in `default.env` because their value needs to be computed from other values as you've said it yourself (ex: `PAVICS_FQDN_PUBLIC`).  But it's the same "have default value" spirit here.

They are also _required_ because they are used in template expansion, example `PAVICS_FQDN_PUBLIC`: https://github.com/bird-house/birdhouse-deploy/blob/57a320c3583f804838e8c18aeb064266527a0bc9/birdhouse/config/canarie-api/docker_configuration.py.template#L7

So what is "incorrect" other than the technicality where the default value comes from?

File: birdhouse/pavics-compose.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/pavics-compose.sh:30]: > But if they don't exist at all (as it can be the case when in `OPTIONAL_VARS`)

So this is exactly why I am confused.  Since the value for those new vars are defined in `default.env` how can they don't exist at all?

Are you trying to defend against git checkout failure due to disk full or hard drive failure? Or user somehow manually editing `default.env` and delete those values knowingly?  If yes, I do not think this is the location to defend against those kind of error since those kind of errors can happen to any code anywhere, not just these variables.

File: birdhouse/env.local.example [hunk 12]
```diff
@@ -17,7 +20,6 @@ export TWITCHER_PROTECTED_PATH=/twitcher/ows/proxy
 export PHOENIX_PASSWORD=phoenix_pass
 export PHOENIX_PASSWORD_HASH=sha256:123456789012:1234567890123456789012345678901234567890123456789012345678901234
 export TOMCAT_NCWMS_PASSWORD=ncwmspass
-export SUPPORT_EMAIL=helpdesk@example.com
```
Comment (tlvu) [inline birdhouse/env.local.example:20]: Can you restore this `SUPPORT_EMAIL` here since it was previously moved to `default.env` and then you reverted that move (agree with the revert by the way, thanks).

File: birdhouse/pavics-compose.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/pavics-compose.sh:30]: For backward compat, please define `DATA_PERSIST_ROOT=/data` in `default.env` and move this `DATA_PERSIST_ROOT` into the `OPTIONAL_VARS` list since it has default value.

Without this backward compat, existing prod deployment that already use `/data` will not start because `DATA_PERSIST_ROOT` is missing in their `env.local` but they do not need to override it since they are already ok using `/data`.

Did the pipeline test worked in this state that `DATA_PERSIST_ROOT` is not in `default.env` and not in `env.local` but is in mandatory `VARS` list?

**Review by tlvu — CHANGES_REQUESTED** [review]
Almost there.  2 items in this review below plus
* mongodb volume-mount (you either forgot to use it in this PR or if really not needed in this PR, then remove it here and add it when really needed)
* remove duplicate default value (`/data`) definition in `deploy_data_job.env`

**Comment by tlvu** [comment]
I have also updated the PR description with the following to summarize this PR:

=======
Change overview:
* allow to customize `/data` persistence root on disk, retaining current default for existing deployment
* add data persistence for `mongodb` container
* secure mongodb access by putting the `mongodb` container into its own network, separating it from the `default` network of all the other containers in the stack.  Going forward, if a container need to communicate with mongodb, it needs to also be in the same `mongodb` network.

=======

Please copy/paste the PR description into the merge description if you do the merge so we do not depend on github for this summary.

File: birdhouse/pavics-compose.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/pavics-compose.sh]: I think that in a configuration point of view, for an external dev configuring a new platform, a variable that defaults to some value if not provided, regardless of how that resolution occurs, is not required. 
It would be even more misleading to say that `OPTIONAL_VARS` is required.

File: birdhouse/pavics-compose.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/pavics-compose.sh:30]: The point is that they must exist, not that they won't possibility exist because of defaults. 

If for some reason `DATA_PERSIST_ROOT` gets renamed or removed at some point, having them in `OPTIONAL_VARS` will silently pass, and then the script will raise some permission error when trying to create dirs. 
Worst, it gets executed as root and pass completely silently and things break without much harder investigations to find the cause. 

Having the value in `VARS` forces `pavics-compose` to ensure they are found and *defined*. Imagine someone incorrectly overrides `env.local` with their custom variable `DATA_PERSIST_ROOT="${SOME_MISSING_VAR}"`, but because of bad order in sourcing files, it resolves as `""`. I think the explicit error is much better and will help anyone quickly identify the causes.

File: birdhouse/pavics-compose.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/pavics-compose.sh]: > I think that in a configuration point of view, for an external dev configuring a new platform, a variable that defaults to some value if not provided, regardless of how that resolution occurs, is not required.

Yes, "not required" to be set in `env.local`, but not "not required" for operating the product.  If a var is not required for operating the product, then it would not exist at all.

> It would be even more misleading to say that OPTIONAL_VARS is required.

It's "optional" for `env.local`.  Just to be clear, it has always been about their presence in the `env.local` file.

**Comment by fmigneault** [comment]
I would like to avoid managing yet another PR just for mongo, considering how small it is. 
I have 5 pending PR right now, all inter-dependant, and I'm wasting more time syncing them than anything else.

I think that `/data` should still be as explicit default  in the non-template files, because there is nothing that ensures both `default.env` and `env.local` get source'd before their execution.

File: birdhouse/pavics-compose.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/pavics-compose.sh:30]: > If for some reason `DATA_PERSIST_ROOT` gets renamed or removed at some point, having them in `OPTIONAL_VARS` will silently pass, and then the script will raise some permission error when trying to create dirs.
> Worst, it gets executed as root and pass completely silently and things break without much harder investigations to find the cause.
> 
> Having the value in `VARS` forces `pavics-compose` to ensure they are found and _defined_. Imagine someone incorrectly overrides `env.local` with their custom variable `DATA_PERSIST_ROOT="${SOME_MISSING_VAR}"`, but because of bad order in sourcing files, it resolves as `""`. I think the explicit error is much better and will help anyone quickly identify the causes.

Okay but these problem of accidental rename a var (this must somehow pass PR review) or accidental overriding the default value to empty value in `env.local` can happen to every variables.  Are you suggesting putting all vars from `OPTIONAL_VARS` and all vars in `default.env` into `VARS` as well so they are checked for not empty value?

What if the value is not empty but not valid?

Where do we stop?

I am not against making the code more error resilient and robust.  But it's probably outside the of scope of this PR.  Let's keep the consistency in this PR and we can discuss this issue somewhere else.

File: birdhouse/env.local.example (hunk 12 above)
Comment (fmigneault) [reply birdhouse/env.local.example]: done

File: birdhouse/pavics-compose.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/pavics-compose.sh:30]: Not every var, but just the ones that require to be explicitly checked for undefined or empty.
Some variables are valid as empty (eg: not used feature), those should remain in `OPTIONAL_VARS`. 

`DATA_PERSIST_ROOT` is a special use case that cannot be empty, but can use a default if omitted.
So explicit undefined is ok, empty is not. This is what I meant originally that there are 3 states, and only 2 are checked by `VARS` and `OPTIONAL_VARS` (by themselves). Adding the variable in `VAR` + in `default.env` fulfills that 3rd state of undefined OK but not empty via default.

In fact, the real way to fix it *should* be to make it required, but is not for backward compatibility.

[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/pavics-compose.sh:30]: I hear you.  We can make the code more robust and error resilient but in a different issue to avoid dragging this PR further.

`VARS` and `OPTIONAL_VARS` was for really initially for templating use and we want the list to be as minimal as possible to avoid template expanding a var that was not meant to be.

For your needs, I see maybe a 3rd var `CHECK_NOT_EMPTY` that will contain all vars in `default.env` that should not be empty so we do not hijack `VARS` for the wrong intended use.  Some auditing involved as not every var is invalid if empty.

**Comment by tlvu** [comment]
> I would like to avoid managing yet another PR just for mongo, considering how small it is.

I quickly look at your Weaver PR and I do not see where this mongodb volume-mount is being used. https://github.com/bird-house/birdhouse-deploy/blob/69e983178287b4a4843b93c1aae05c185eb9388e/birdhouse/components/weaver/docker-compose-extra.yml#L48-L54

I just do not want to create un-used code from the start.

On another note, FYI an optional component can also create it's own data volume, example https://github.com/bird-house/birdhouse-deploy/blob/02d1e9db0cfc0ca0c2ca382e379228a48d4ca182/birdhouse/optional-components/testthredds/docker-compose-extra.yml#L45-L47 so if you need it for Weaver, it does not have to be another small PR because the Weaver PR can do it but should probably be named `mongodb_weaver` to be clear that it is "for and from" weaver component.

> I think that `/data` should still be as explicit default in the non-template files, because there is nothing that ensures both `default.env` and `env.local` get source'd before their execution.

I am guessing this reply is for `deploy_data_job.env`.  This is sourced by `env.local` and `env.local` is sourced by `pavics-compose.sh` which ensure `default.env` is sourced before `env.local`.  So this is a far as the code can do to re-assure you the sourcing order.

I understand you want error resilient code but duplicating/re-defining default value is also a source for possible errors.

This has been a long PR.  I do not want to drag it more so if it makes you sleep better at night to duplicate default value again I'll let go this one.

**Comment by fmigneault** [comment]
@tlvu 
> I quickly look at your Weaver PR and I do not see where this mongodb volume-mount is being used. 

The line directly above your linked segments, indirectly via network. 
Since mongo is included in the "base" docker-compose, if believe the network and volume should be defined there also. 
I agree that with the 2 PR, it is harder to track, so I can add it via Weaver feature. 

Nice, for `deploy_data_job.env`, then should be safe to remove in that case. 
I am still not sure about the other ones in `backup-jupyterhub-notebooks.sh` and `bootstrap-testdata`. 

Just to be clear, I'm not adding those defaults for myself (I don't use those scripts specifically), but rather to ensure existing servers don't break if sourcing doesn't occur accordingly. If you feel this is safe enough to remove it, I will, but I can't confirm their result. I only know that the default/env.local via `pavics-compose.sh` works because I override the root for persistence volumes on my side.

File: birdhouse/pavics-compose.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (fmigneault) [reply birdhouse/pavics-compose.sh:30]: A new list for the 3rd case might not look like a big difference here, but it does impact more when reaching weaver feature. 

I need to dynamically extend the `VARS` here: 
https://github.com/bird-house/birdhouse-deploy/blob/feature/DAC-300-weaver-compose/birdhouse/components/weaver/default.env#L5-L21

The reason I need to work with `VARS`, is that I also use `DATA_PERSIST_ROOT`, which then generates paths to volumes for data-persistence, which them MUST exist. 
This dynamic "hack" of `VARS` felt already unusual enough as is, which is why I initially didn't introduce another list of vars that increases even more the complexity of weaver feature. 

But sure, a subsequent PR after all that could be considered.

**Comment by tlvu** [comment]
> The line directly above your linked segments, indirectly via network.
> Since mongo is included in the "base" docker-compose, if believe the network and volume should be defined there also.

Are you implying since you define a custom network for mongodb, you also need a matching data volume name?  I don't think they need to be kept related.

I think the network lock down is great and should even stay in the base to promote security for all future users of mongodb container.

The data volume to share data at file level between mongodb and another container is more of a customization and should not even be shared.  You wouldn't want another container to access to files that Weaver want to share with mongodb !

But you already removed the network for now.  We can keep it and add it back later if needed.  This PR is long enough.

> Nice, for `deploy_data_job.env`, then should be safe to remove in that case.
> I am still not sure about the other ones in `backup-jupyterhub-notebooks.sh` and `bootstrap-testdata`.

I already thumbs up for `backup-jupyterhub-notebooks.sh` and `bootstrap-testdata` to duplicate the `/data` default value since they do not source any `env.local` or `default.env`.

File: birdhouse/pavics-compose.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply birdhouse/pavics-compose.sh:30]: > I need to dynamically extend the `VARS` here:
> https://github.com/bird-house/birdhouse-deploy/blob/feature/DAC-300-weaver-compose/birdhouse/components/weaver/default.env#L5-L21

Well, the 3rd new var `CHECK_NOT_EMPTY` will have be extensible by components, just like `VARS` because each component will need to inject their own new variables.  Each component will have to be able say here are my new variables that should not be empty (add to `CHECK_NOT_EMPTY`) and a subset of those new variables will be used for template expansion (add to `VARS`).

If you worry how to extend that 3rd new var `CHECK_NOT_EMPTY`, no worries !

**Review by tlvu — APPROVED** [review]
I am okay with this PR for now.  I can let you hijack `VARS` for the wrong purpose for now and delay implement the proper `CHECK_NOT_EMPTY` in another PR, to not block this PR.

But let's agree to not hijack `VARS` more than it is now.  `VARS` has been meant for template expansion and not for ensuring vars not empty.  We do not want to accidentally template expand a var that was not meant to be.

Just to be sure, an automated pipeline has started with this PR and Jenkins errors are all known?  I have not attempted to manually verify this PR.  Just to make sure no typo errors.

**Comment by fmigneault** [comment]
@tlvu Yes. All remaining errors are only the ones related to the single failing magpie/thredds test and the finch subset tests.

**Comment by tlvu** [comment]
> @tlvu Yes. All remaining errors are only the ones related to the single failing magpie/thredds test and the finch subset tests.

OK, merge when ready.  Remember to copy/paste PR description into merge description.
